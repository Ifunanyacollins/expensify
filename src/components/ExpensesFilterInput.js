import React from 'react';
import { connect } from 'react-redux';
import {  filterByText,sortByAmount,sortByDate, setStartDate, setEndDate } from '../actions/filters'
import 'react-dates/initialize'
import {DateRangePicker} from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'


export class ExpensesFilterInput extends React.Component{

  constructor(props){
    super(props)
    this.state = {
      focusedInput:null
    }
  }

 onDatesChange = ({startDate,endDate}) =>{
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
 }

 onFocusChange = (focusedInput) =>{
  this.setState(()=>({focusedInput}))
 }

 onTextchange = (e)=>{
  this.props.filterByText(e.target.value)
 }
 
 onSortBy = (e)=>{
  (e.target.value && e.target.value === 'date') ? this.props.sortByDate() : this.props.sortByAmount() 
}

  render(){
    
    return(
      <div>

     <input type="text" value={this.props.filters.text} onChange={this.onTextchange}/>

     <select value={this.props.filters.sortBy} onChange={this.onSortBy}>
     <option value="date">Date</option>
     <option value="amount">Amount</option>
     </select>

      <DateRangePicker 
       startDate={this.props.filters.startDate}
       startDateId={'1'}
       endDate={this.props.filters.endDate}
       endDateId={'2'}
       onDatesChange={this.onDatesChange}
       focusedInput={this.state.focusedInput}
       onFocusChange={this.onFocusChange}
       numberOfMonths={1}
       isOutsideRange={()=>false}
       showClearDates={true}

      />
      </div>
   )


  }
}

const mapDispatchToProps = (dispatch,props)=>({
  setStartDate:(startDate)=>dispatch( setStartDate(startDate) ),
  setEndDate:(endDate)=>dispatch( setEndDate(endDate)),
  filterByText:(text)=>dispatch( filterByText(text)),
  sortByDate:()=>dispatch( sortByDate() ),
  sortByAmount:()=>dispatch( sortByAmount() )
})


const MapToProps = ({expenses,filters}) => {

  return{
      filters
  }

}


export default connect(MapToProps,mapDispatchToProps)(ExpensesFilterInput)