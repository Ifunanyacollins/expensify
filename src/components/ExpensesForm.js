import React from 'react'
import moment from 'moment'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'





export default class ExpensesForm extends React.Component{

        constructor(props){
            super(props);

            this.state = {
                description:props.expense ? props.expense.description : '',
                note:props.expense ? props.expense.note : '',
                amount:props.expense ? props.expense.amount.toString() : '',
                createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
                focused:false,
                error:undefined
            }
        }
    
    



     onDescriptionChange= (e)=>{

         const value = e.target.value
         this.setState(()=>({description:value}))
       
     }
     
     onNoteChange = (e) => {
         const value = e.target.value
         this.setState(()=>({
             note:value
         }))
     }

     onAmountChange = (e) => {
         const value = e.target.value
         if(!value || value.match(/^\d{1,}(\.\d{0,2})?$/))
         this.setState(()=>({amount:value}))
        }

        onDateChange = (createdAt)=> {
            if(createdAt){
                this.setState(()=>({createdAt}))
            }
          
          
        }

        onFocusChange = ({ focused }) =>{
            this.setState(()=>({focused}))
           
        }
        onFormSubmit = (e) =>{
            e.preventDefault()

            if(!this.state.description || !this.state.amount){
               this.setState(()=>({error:true})) 
            }else{
                this.setState(()=>({error:undefined}))
                this.props.onSubmit({
                    description:this.state.description,
                    amount:parseFloat(this.state.amount,10),
                    createdAt:this.state.createdAt.valueOf(),
                    note:this.state.note
                })
            }
        }

    render(){

        return(
            <div>
               {this.state.error && <p>Hmmm, it seems you didnt provide a description or Amount</p>}
               <form onSubmit={this.onFormSubmit}>
               <input 
                type="text"
                placeholder="description" 
                value={this.state.description}
                onChange={this.onDescriptionChange}
                />
               <input type="text" placeholder="amount" value={this.state.amount} onChange={this.onAmountChange} />
               <SingleDatePicker

                date={this.state.createdAt}
                onDateChange={this.onDateChange}
                focused={this.state.focused}
                onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={()=>false}
                
               />
               <textarea value={this.state.note} onChange={this.onNoteChange}></textarea>
               <button>Add Expenses</button>
               </form>
            </div>
        )

    }
}