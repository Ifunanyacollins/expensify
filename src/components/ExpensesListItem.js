import React from 'react';
import {Link} from 'react-router-dom';


const ExpensesListItem = ({description,amount,createdAt,note ,id})=>{
 return(
     <div>
      <p>Title of Payment = >
      <Link to={`/edit/${id}`} >{description}</Link>
      </p>
      <p> Amount = > {amount}</p>
      <p>Time & Date = >{createdAt}</p>
      <p>Note = > {note}</p>
      <br/>
     </div>
 )
}

export default ExpensesListItem;