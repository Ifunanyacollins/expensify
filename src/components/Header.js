import React from 'react'
import {NavLink} from 'react-router-dom';
import '../styles/components/Header.css'

 const Header = () =>(
    <header>
    <h4>Expensify</h4>
    <ul>
    
    <li><NavLink to="/" activeClassName="is-active" exact={true}>Home</NavLink></li>
    <li><NavLink to="/create" activeClassName="is-active">Add New</NavLink></li>
    <li><NavLink to="/dashboard" activeClassName="is-active">Dashboard</NavLink></li>
    <li><NavLink to="/help" activeClassName="is-active">Help</NavLink></li>
    
    </ul>
    </header>
);

export default Header;