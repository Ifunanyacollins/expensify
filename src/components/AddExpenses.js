import React from 'react';
import { connect } from 'react-redux';
import {add_expense} from '../actions/expenses'
import ExpensesForm from './ExpensesForm';


 export class AddExpenses extends React.Component{
 onSubmit = (expense)=>{
     this.props.add_expense(expense)
     this.props.history.push('/dashboard')
 }
    render(){
        return(
            <ExpensesForm  onSubmit={this.onSubmit}/>
        )
    }
}




const mapDispatchToProps = (dispatch) => ( {
    add_expense:(expense) => dispatch(add_expense(expense))
} )


export default connect(undefined,mapDispatchToProps)(AddExpenses);