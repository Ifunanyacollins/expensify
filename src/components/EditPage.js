import React from  'react'
import { connect } from 'react-redux'
import {edit_expense,remove_expense } from '../actions/expenses'
import ExpensesForm from './ExpensesForm'



 export class EditExpenses extends React.Component{

  onSubmit = (expense)=>{
    this.props.edit_expense(this.props.expense.id , expense)
    this.props.history.push('/dashboard')
  }

  onClickhandle = ()=>{
    this.props.remove_expense(this.props.expense.id)
    this.props.history.push('/dashboard')
  }

  render(){
    return(
      <div>
      <ExpensesForm 
       expense={this.props.expense}  
       onSubmit={this.onSubmit}
      />
      <button
      onClick={this.onClickhandle} 
      >Remove
      </button>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch ,props)=>(
  {
   edit_expense:(id,expense)=>( dispatch(edit_expense(id,expense)) ),
   remove_expense:(id)=>( dispatch(remove_expense(id)) )
  }
)

const MapToProps = (state,props)=>{
  
  return{ expense:state.expenses.find((expense)=>(
     expense.id === props.match.params.id
  )) }
}


export default connect(MapToProps,mapDispatchToProps)(EditExpenses);