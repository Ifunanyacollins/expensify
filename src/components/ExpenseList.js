import React from 'react';
import {connect} from 'react-redux';
import ExpensesListItem from './ExpensesListItem';
import Selectors from '../selectors/expenses';
import ExpensesFilterInput from './ExpensesFilterInput';


export const ExpenseList = (props)=>{

  return(   
 <div>
 <ExpensesFilterInput />
 {
     props.expenses.length === 0 ? (
       <p>No Expenses</p>
     ):
     (
        props.expenses.map((item,index)=><ExpensesListItem {...item} key={index}/>)
     )
 }
  
 </div>
)

}


const MaptoState = ({expenses,filters})=>(
    {
        expenses:Selectors(expenses,filters)
    }
);

export default connect(MaptoState)(ExpenseList);