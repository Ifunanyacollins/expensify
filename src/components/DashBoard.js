import React from 'react';
import ExpenseList from './ExpenseList'


 const DashBoard = () => (
  <div>
    <p>Hello Welcome to Dashboard</p>
    <ExpenseList />
  </div>  
);

export default DashBoard;