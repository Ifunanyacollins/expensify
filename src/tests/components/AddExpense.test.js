import React from 'react';
import {shallow} from 'enzyme';
import { AddExpenses } from '../../components/AddExpenses'
import {expenses} from '../fixture'


let add_expense,history,wrapper;

beforeEach(()=>{
     add_expense = jest.fn()
     history = {push:jest.fn()}
     wrapper = shallow(<AddExpenses  add_expense={add_expense} history={history}/>)
})

test('render addExpenses page',()=>{
    expect(wrapper).toMatchSnapshot()
})

test('render addExpenses page with value',()=>{
    wrapper.find('ExpensesForm').prop('onSubmit')(expenses[1])
    expect(history.push).toHaveBeenLastCalledWith('/dashboard')
    expect(add_expense).toHaveBeenLastCalledWith(expenses[1])
    expect(wrapper).toMatchSnapshot()
})