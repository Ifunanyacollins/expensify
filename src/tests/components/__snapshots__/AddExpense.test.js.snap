// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render addExpenses page 1`] = `
<ExpensesForm
  onSubmit={[Function]}
/>
`;

exports[`render addExpenses page with value 1`] = `
<ExpensesForm
  onSubmit={[Function]}
/>
`;
